### DISCUSSION TITLE
Nested custom adapter errors with createEntityAdapter

### DISCUSSION THREAD

**Discussion by pBouillon** [state-adapt/state-adapt#75] [Q&A]
Hi,

I would like to implement the following structure for my state:

- A `Game` has some `Player`s
- A `Player` has a `Score`
- A `Score` is made of `points` and a `missesStreak`

The breakdown of how I did this state with state-adapt is the following one:

### Score

```ts
export const pointsAdapter = createAdapter<number>()({
  update: (score, fallenPins: Pin[]) => fallenPins.length === 1 ? fallenPins[0] : fallenPins.length,
});

export const missesStreakAdapter = createAdapter<number>()({
  update: (streak, fallenPins: Pin[]) => fallenPins.length > 0 ? 0 : streak + 1,
});

export const scoreAdapter = joinAdapters<Score>()({
  currentMissesStreak: missesStreakAdapter,
  points: pointsAdapter,
})({
  addFallenPins: {
    points: pointsAdapter.update,
    currentMissesStreak: missesStreakAdapter.update,
  },
})();
```

### Player

```ts
export const playerAdapter = joinAdapters<Player>()({
  id: createAdapter<string>()({}),
  name: stringAdapter,
  score: scoreAdapter,
})({
  hasWon: (state: Player) => state.score.points === 50,
})();

export const playerEntityAdapter = createEntityAdapter<Player, PlayerId>()(
  playerAdapter,
  { filters: ['hasWon'] },
);
```

### Game

```ts
export const initialState: Readonly<GameState> = {
  players: createEntityState<Player, PlayerId, GameState['players']>(),
  currentPlayerIndex: 0,
};

export const gameAdapter = joinAdapters<GameState>()({
  players: playerEntityAdapter,
  currentPlayerIndex: numberAdapter,
})({
  nextPlayerIndex: ({ currentPlayerIndex, playersAll }): number => (currentPlayerIndex + 1) % playersAll.length,
  currentPlayer: ({ currentPlayerIndex, playersAll }) => playersAll.at(currentPlayerIndex),
  winner: (state): Player | null => state.playersHasWon.at(0) || null,
})();
```

I would have hoped to take advantage of state-adapt composition's API to be able to update the state of a `Player` from an array of `Pin` directly from the `gameAdapter`

Also, I did this to that end:

```ts
  readonly #watchedAdapter = watch(this.#path, gameAdapter);

  readonly endCurrentRound$ = new Source<{ fallenPins: Pin[] }>(
    '[Game] End Current Round',
  );

  readonly #addCurrentPlayerFallenPins$ = this.endCurrentRound$.pipe(
    switchMap(({ payload }) =>
      this.#watchedAdapter.currentPlayer$.pipe(
        take(1),
        map(({ id }: Player): [string, Pin[]] => [id, payload.fallenPins]),
      ),
    ),
    toSource('[Game] addCurrentPlayerFallenPins$'),
  );

  readonly #store = adapt(initialState, {
    path: this.#path,
    adapter: gameAdapter,
    sources: {
      // ...
      addPlayersOneScoreFallenPins: this.#addCurrentPlayerFallenPins$,
  });
```

This code compiles fine but, when `#addCurrentPlayerFallenPins$` emits, I observe the following behavior:

1. The store is destroyed
  ![image](https://github.com/state-adapt/state-adapt/assets/22640284/70fd6412-322a-4058-9979-1bc46fdf6412)

2. Several errors are thrown
  ![image](https://github.com/state-adapt/state-adapt/assets/22640284/32c8efc5-d57c-44d2-af9c-33b3f59705ea)

I suspected `scoreAdapter` to be incorrect so I created a state only from it and called its update with array of `Pin` and I observed the desired outcome (the score and the streak correctly updating). I also reviewed the other selector's behaviors and they seem to respond properly on the other parts of the application.

Is there something I missed in the usage of a custom adapter from within the `createEntityAdapter`? Did I do wrong elsewhere?

Thanks a lot for your help! This might be a bit long but I'm really confused about what is the root of the error so I tried to include as many details as possible for the parts that might be concerned.

**Comment by pBouillon** [comment]
**Update:** I tried also using the player directly:

```ts
test$ = new Source<Pin[]>('Test');

#state = adapt(emptyPlayer, {
  path: 'player',
  adapter: playerAdapter,
  sources: {
    addScoreFallenPins: this.test$,
  },
});

state = toSignal(this.#state.state$);
```

and displaying the result in the template:

```html
<pre>{{ state() | json }}</pre>
<button (click)="test$.next([1, 2])">Trigger</button>
```

And it works great, same for the Redux Devtools insights

It really feels like this comes from `createEntityAdapter`. Do you have any other lead on this point?

**Comment by mfp22** [comment]
Your code looks good to me. 

One difference with the way I do things is 
```typescript
nextPlayerIndex: ({ currentPlayerIndex, playersAll }): number => (currentPlayerIndex + 1) % playersAll.length,
```
instead of 
```typescript
nextPlayerIndex: (s): number => (s.currentPlayerIndex + 1) % s.playersAll.length,
```
The reason I do it this way is because `s` is a proxy and it waits until something is accessed before computing it. This makes it more efficient than `reselect` (traditional selectors). 

I doubt that's the issue though, so I will copy your code into StackBlitz and see if I can get it to work.

**Reply by mfp22** [reply]
I made this in StackBlitz and I don't see anything bad happening. After 3 seconds I manually call `addCurrentPlayerFallenPins$` and  I get no errors. https://stackblitz.com/edit/state-adapt-angular-cenkip?file=src%2Fapp%2Fapp.component.ts,tsconfig.json

What versions of everything are you using, including TypeScript? 

It would help a lot if you could reproduce it in StackBlitz.

**Reply by pBouillon** [reply]
Thanks for the follow-up, I really appreciate it

I'm using Angular 17.1, StateAdapt 2.0.3 and TS 5.2.2

I made a small update to the StackBlitz, and I think that I reproduce the same error I encountered.
I suspect it being linked to the way the players are added.

Here is the updated environment in which two players are created on initialization, and you can end the current round on clicking the button: https://stackblitz.com/edit/state-adapt-angular-bwbuc2?file=src%2Fapp%2Fapp.component.ts

On click, the store destroys itself and the console is showing several errors.

**Reply by mfp22** [reply]
Thanks.

It's most likely a problem with the entity adapter, but it's hard to debug in StackBlitz. I'm going to create an issue and reference this. 

The entity adapter is the least tested part of StateAdapt. It's probably as well battle tested as most of the rest of StateAdapt was in 2022.

I'm going to compare the differences between my repro and yours and see if I can figure it out. Otherwise I will have to debug it locally. I'll add a unit test either way.

**Reply by pBouillon** [reply]
Glad it narrowed down the issue!

If I may suggest, the entity adapter could benefit from a more detailed documentation as well, it was a bit hard getting into it.

However, this is not directly related to the issue, I just wasn't sure if this was a bug or me not correctly following the correct usage.

A small tutorial on how to use it, like NgRx entities for example, could help people getting into it more easily.

**Reply by mfp22** [reply]
The way to do this without an error is to calculate the correct initial state using the adapter directly:

```typescript
const emptyGameState: GameState = {
  players: createEntityState<Player, PlayerId, GameState['players']>(),
  currentPlayerIndex: 0,
}

export const initialState = gameAdapter.setPlayersAll(emptyGameState, [
  { id: '1', name: 'John', score: { ...emptyScore } },
  { id: '2', name: 'Jane', score: { ...emptyScore } },
], emptyGameState);
```

instead of having a source emit an event to set it in the store:

```typescript
  readonly addPlayers$ = new Source<void>('addPlayers$').pipe(
    startWith({}),
    map((): Player[] => [
      { id: '1', name: 'John', score: { ...emptyScore } },
      { id: '2', name: 'Jane', score: { ...emptyScore } },
    ]),
    toSource('createPlayers$')
  );

// ...

      setPlayersAll: this.addPlayers$,
```

I'm still trying to figure out where this error is exactly coming from, but this is why I hadn't seen it before. I don't like sources emitting unless there was a real event that caused it. It's simpler to calculate an initial state than to use state changes. 

But the real problem is that I don't explain this anywhere. You're right, I definitely need better documentation.

**Reply by pBouillon** [reply]
The issue arrises in my case where a new game starts.

At this point, the user is shown a form and can type as many names as he wants. Those names will then be the one emitted as the creation of a game based on those names, as an event of "hey, a new game will start, with those players" and some other sources might also react to that (cleaning current state, etc.)

If that's not possible I can try to achieve that by another means, it just doesn't look like an edge case so it felt strange to run into that bug.

By any chance, did you had the time to check #81? It might be an alternative way to work on collections that could be more simple.

**Reply by mfp22** [reply]
Right, I see. So the problem isn't the synchronous `setPlayersAll` but the fact that it wasn't part of the initial state at all. I bet the selectors cache isn't being created properly. I will debug this locally now.

**Reply by mfp22** [reply]
I figured it out.

In order to call state change functions for individual entities, we need to supply all the arguments `state, payload, initialState, cache`. But when we do something to a single entity, that `initialEntityState` may be `undefined` because it was technically added later so the original the whole `initialState` didn't include it. This means the inner adapter's state change function might receive `undefined` as its `initialState` argument. 

In this case, the inner adapter was created with `joinAdapters`, which means for every single function, it passes `initialState[namespace]` down to its own inner adapters, so `undefined[namespace]` was erroring. 

I could fix `joinAdapters` to handle this. I can't think of any state change functions that really need `initialState` other than `reset`. There could be some. But the entity adapter does have `resetOne` and `resetMany`, and if `initialState` returns `undefined` for those, it can just perform the same function as `removeOne` or `removeMany` (set to `undefined` and remove from `ids`). I don't think it makes sense to redefine `initialState` to some special later state. 

So, adapters passed to `createEntityAdapter` should be able to tolerate `undefined` as `initialState`. I'll think about how this could maybe be handled with TypeScript... but I think it's rare outside of a special passthrough function like `joinAdapters`. 

If I type the 3rd argument to state change functions `initialState` as optional, it cascades to a lot of annoying TypeScript work for everyone. This could be worth handling...

I'm trying to think of at least a helpful error or warning to add, but I can't. For all entities that are added later, it would be logging a warning for every single inner state change function every time it ran. 

Lastly, the entity or array adapters could have documentation that warns about this corner case. 

But for now I'm adding a fix for `joinAdapters`.

**Reply by mfp22** [reply]
I finally released 2.0.6 with this fix, but have found the same kind of problem inside `buildAdapter` with group reactions. I'll release that fix with 2.0.7 tomorrow. But in 2.0.6 the workaround is to use plain reactions instead of group reactions, as shown here https://stackblitz.com/edit/state-adapt-angular-voaeqz?file=src%2Fapp%2Fapp.component.ts

**Comment by mfp22** [comment]
Alright, 2.0.7 is released and this is working in StackBlitz https://stackblitz.com/edit/state-adapt-angular-voaeqz?file=src%2Fapp%2Fapp.component.ts
